src/ray/raylet/reconstruction_policy.cc

```diff
@@ -4,8 +4,146 @@ namespace ray {
 
 namespace raylet {
 
-void ReconstructionPolicy::CheckObjectReconstruction(const ObjectID &object) {
-  throw std::runtime_error("Method not implemented");
+ReconstructionPolicy::ReconstructionPolicy(
+    boost::asio::io_service &io_service,
+    std::function<void(const TaskID &)> reconstruction_handler,
+    int64_t initial_reconstruction_timeout_ms, const ClientID &client_id,
+    gcs::PubsubInterface<TaskID> &task_lease_pubsub,
+    std::shared_ptr<ObjectDirectoryInterface> object_directory)
+    : io_service_(io_service),
+      reconstruction_handler_(reconstruction_handler),
+      initial_reconstruction_timeout_ms_(initial_reconstruction_timeout_ms),
+      client_id_(client_id),
+      task_lease_pubsub_(task_lease_pubsub),
+      object_directory_(std::move(object_directory)) {}
+
+void ReconstructionPolicy::SetTaskTimeout(
+    std::unordered_map<TaskID, ReconstructionTask>::iterator task_it,
+    int64_t timeout_ms) {
+  task_it->second.expires_at = current_sys_time_ms() + timeout_ms;
+  auto timeout = boost::posix_time::milliseconds(timeout_ms);
+  task_it->second.reconstruction_timer->expires_from_now(timeout);
+  const TaskID task_id = task_it->first;
+  task_it->second.reconstruction_timer->async_wait(
+      [this, task_id](const boost::system::error_code &error) {
+        if (!error) {
+          auto it = listening_tasks_.find(task_id);
+          RAY_CHECK(it != listening_tasks_.end());
+
+          if (it->second.subscribed) {
+            // We did not receive a task lease notification within the lease period.
+            // The lease is expired, so attempt to reconstruct the task.
+            HandleTaskLeaseExpired(task_id);
+          } else {
+            // This task is still required, so subscribe to task lease notifications.
```

Comment:
Yes, I think you have it right. I added this branch instead of subscribing to the task lease table immediately in case the caller decides that the object is no longer needed within the timeout. I'll clarify the comment.